Port to 1.20.1

ToCraft opened the pull request:
I've split it into multiple commits so you can check each commit sperately.
There are still some compiler errors, but I think you can already take a look at the current state :smiley: 

---

I think at least the following is worth of discussing:
- I've replaced the Materials with BlockBehaviours, e.g. Material.STONE with Properties.copy(Blocks.STONE)
- Sometimes I've used BuiltInRegistries and ForgeRegistries, but Forge wants modders to only use ForgeRegistries when possible afaik
- rn CreativeTabs collect their items by adding the item to a list when it's registered (see [addToTabItemList(item, list)](https://github.com/ToCraft/TinkersConstruct/blob/1643d8bb5c25ec971387b428cb2e1c7d7e910708/src/main/java/slimeknights/tconstruct/common/TinkerModule.java#L96) )
- rn The Tag Providers get a list of Elements and get their keys by iterating. Maybe there's a more direct way to get the ids (e.g. [getBlocksResourceKeys(blocks)](https://github.com/ToCraft/TinkersConstruct/blob/6258daf9d8b0ad15bb4add05e38ca186c53d1210/src/main/java/slimeknights/tconstruct/common/data/tags/BlockTagProvider.java#L651) )
- I've stripped the "getName" method in many DataProviders since it's a final method now, name should be set somehow else
- The DataProvider's run method can now return a CompleteableFuture, maybe some stuff should therefore be run asynchronously?

---

Some of the things I still haven't taken a look at:
- Recipes
- Modifiers
- DamageSources

ToCraft committed:
update gradle

ToCraft committed:
re-start porting to 1.20.1
- move to new Registries

ToCraft committed:
start replacing materials with BlockBehaviours

ToCraft committed:
some more adjustments to materials

ToCraft committed:
merge to new CreativeTabBuilder

ToCraft committed:
background stuff, start using GuiGraphics, Fix sendSuccess in commands

ToCraft committed:
some more basic stuff

ToCraft committed:
fix some tag providers

ToCraft committed:
fix TransformType

ToCraft committed:
some basic refactoring

ToCraft commented:
I don't know why, but GitHub just closed the PR after I've tried to rename the branch (I'm used to GitHub  auto-updating every reference, seems like it didn't work this time)

ToCraft committed:
simple stuff like level -> level()

KnightMiner commented:
> Sometimes I've used BuiltInRegistries and ForgeRegistries, but Forge wants modders to only use ForgeRegistries when possible afaik


Forge Registries are deprecated, they are removed in 1.21 Neo. I know Forge claims vanilla registries are deprecated, but thats only because the Forge team had no plans to remove them; functionally there has never been a reason to not use the vanilla ones.

> rn CreativeTabs collect their items by adding the item to a list when it's registered (see [addToTabItemList(item, list)](https://github.com/ToCraft/TinkersConstruct/blob/1643d8bb5c25ec971387b428cb2e1c7d7e910708/src/main/java/slimeknights/tconstruct/common/TinkerModule.java#L96) )

Yeah, this is one of those things that makes it hard to review port PRs, such decisions are ones I need to decide how I want it done rather than the code porting over one by one.


> rn The Tag Providers get a list of Elements and get their keys by iterating. Maybe there's a more direct way to get the ids (e.g. [getBlocksResourceKeys(blocks)](https://github.com/ToCraft/TinkersConstruct/blob/6258daf9d8b0ad15bb4add05e38ca186c53d1210/src/main/java/slimeknights/tconstruct/common/data/tags/BlockTagProvider.java#L651) )


I don't know what you mean by this, but it sounds like the wrong approach. I've been watching Forge and never seen any hacks needed to continue tag datagen.

> The DataProvider's run method can now return a CompleteableFuture, maybe some stuff should therefore be run asynchronously?

Look at how Mantle updated its data generators.

> There are still some compiler errors, but I think you can already take a look at the current state 😃

Its really hard to review it properly with compile errors. It makes sense when I am porting on my own, but not when someone else is porting it as I never know if code is the way it is because it has to be or its just not updated.

ToCraft committed:
replace ForgeRegistries with BuiltInRegistries

ToCraft committed:
Generators/Providers now use CompletableFutures

ToCraft commented:
> Forge Registries are deprecated, they are removed in 1.21 Neo. I know Forge claims vanilla registries are deprecated, but thats only because the Forge team had no plans to remove them; functionally there has never been a reason to not use the vanilla ones.

I've changed it as of f0806b5380a1752287e52d3c35849e92d626284e, but in some places I still had to use ForgeRegistries because I could only find some registries there and some methods require ForgeRegistries instead of vanilla registries.

> Yeah, this is one of those things that makes it hard to review port PRs, such decisions are ones I need to decide how I want it done rather than the code porting over one by one.

You have write access in the PR- feel free to adjust it according to your needs. Can you also take a look at the implementations of `Item.fillItemCategory` and `RetexturedBlockItem`? I'm not sure how you'd like to have it ported.

> I don't know what you mean by this, but it sounds like the wrong approach. I've been watching Forge and never seen any hacks needed to continue tag datagen.

It's not that wrong, it's just in case of the `BlockEntityTypeTagProvider`, you've specified `BlockEntityType[]` but in 1.20.1 `ResourceKey<BlockEntityType>[]` is required. That's why I wrote a function that converts your `BlockEntityType[]` into `ResourceKey<BlockEntityType>[]` using the Registries.

>Look at how Mantle updated its data generators.

Alright, I've adapted it as of e036752cb903314a77556e9880ee8126c8aa01af but I'm not sure whether the callbacks should be run asynchronious, too. (Take a look at [AbstractMaterialRenderInfoProvider.run(cache)](https://github.com/ToCraft/TinkersConstruct/blob/e036752cb903314a77556e9880ee8126c8aa01af/src/main/java/slimeknights/tconstruct/library/client/data/material/AbstractMaterialRenderInfoProvider.java#L61) for details)

> Its really hard to review it properly with compile errors. It makes sense when I am porting on my own, but not when someone else is porting it as I never know if code is the way it is because it has to be or its just not updated.

I thought I would create the PR now so that we can coordinate better. But since it is not 100% fixed yet, it is only a draft PR so far.

queengooborg committed:
Remove RetexturedBlockItem and use new Retextured helper classes

queengooborg committed:
Damage source API has been simplified

This commit is not yet working due to an issue getting the holder for BleedingEffect

queengooborg committed:
Entity.level is not a method

queengooborg committed:
Use ItemStack.isSameItem[SameTags]

queengooborg committed:
A RegistryAccess parameter is required for many crafting-related methods

queengooborg committed:
Fix references

queengooborg committed:
Add recipe categories where needed

queengooborg committed:
Cast from float to int as needed

queengooborg committed:
Revert "Fix references"

This partially reverts commit b4def82f7ee78329584cf06232dff27684101be5.

queengooborg committed:
Fix references using same method as ToCraft

queengooborg committed:
Merge remote-tracking branch 'upstream/1.19.2' into queengooborg

queengooborg committed:
ElementScreen constructor no longer has default texture size

queengooborg committed:
GuiGraphics.blit needs the texture

queengooborg committed:
Various fixes

queengooborg committed:
JsonHelper.parseColor is deprecated and scheduled for removal

queengooborg committed:
Final methods cannot be overridden

queengooborg committed:
Add missing import

queengooborg committed:
ElementScreen needs texture and texture size during construction

queengooborg committed:
Remove unused ElementScreen elements

queengooborg committed:
Fix SlotButtonItem class

queengooborg committed:
Fix BorderWidget.draw()

queengooborg committed:
Various fixes

queengooborg committed:
Remove final method override

ToCraft committed:
replace some PoseStacks with GuiGraphics

ToCraft committed:
replace ModelBakery with ModelBaker

ToCraft committed:
Merge branch 'SlimeKnights:1.19.2' into main

queengooborg committed:
Fix DamageSource reference

queengooborg committed:
Call level() when needed

queengooborg committed:
Cast to int as needed

queengooborg committed:
Remove final overrides

queengooborg committed:
Fix calls to Gui.renderSlot()

queengooborg committed:
Remove final override

queengooborg committed:
Add missing import

queengooborg committed:
Use DamageSources registry

queengooborg committed:
PoseStack matrices -> GuiGraphics graphics

queengooborg committed:
Add crafting recipe categories

queengooborg committed:
draw() no longer returns width

queengooborg committed:
Add one more recipe category

queengooborg committed:
Call level() where needed

queengooborg committed:
Add additional recipe categories

queengooborg committed:
Add the last recipe categories

queengooborg committed:
Fix a previous typo

queengooborg committed:
GuiGraphics matrixStack -> graphics

queengooborg committed:
Various fixes

queengooborg committed:
Use Lombok `@Getter`

queengooborg committed:
Recipes require registry access variable

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/tools/capability/inventory/ToolInventoryCapability.java:
```diff
@@ -274,7 +274,7 @@ public ItemStack insertItem(int slot, ItemStack stack, boolean simulate) {
       // space leftover? does it match?
       int limit = Math.min(current.getMaxStackSize(), slotLimit);
-      if (current.getCount() >= limit || !current.sameItem(stack)) {
+      if (current.getCount() >= limit || !current.is(stack.getItem())) {
```
Are you sure this is equivalent? The is method that takes an item is not new, IIRC this is supposed to be checking the item and NBT.

ToCraft replied:
Yep, because that's the underlying code of `isSameItem` in `ItemStack`.
~~~java
public static boolean isSameItem(ItemStack stack, ItemStack other) {
   return stack.is(other.getItem());
}
~~~

However, there is `isSameItemSameTags`, which also compares the NBT Tags. Maybe that's what we should use instead?
~~~java
public static boolean isSameItemSameTags(ItemStack pStack, ItemStack pOther) {
    if (!pStack.is(pOther.getItem())) {
        return false;
    } else {
        return pStack.isEmpty() && pOther.isEmpty() ? true : Objects.equals(pStack.tag, pOther.tag) && pStack.areCapsCompatible(pOther);
    }
}
~~~

queengooborg replied:
Note: the `sameItem()` method's source code is the following:

```java
   /**
    * Compares the {@code ItemStack} argument to this {@code ItemStack}, returns {@code true} if the {@code Items}
    * contained in both {@code ItemStacks} are equal.
    */
   public boolean sameItem(ItemStack pOther) {
      return !pOther.isEmpty() && this.is(pOther.getItem());
   }
```

So now, we're just calling `is()` directly, but skipping the check for `other.isEmpty()`, which is probably fine since we're already checking the count?

It's also worth noting that the new `isSameItem()` method that replaced `sameItem()` just calls `is()` without calling `isEmpty()`:

```java
    public static boolean isSameItem(ItemStack pStack, ItemStack pOther) {
        return pStack.is(pOther.getItem());
    }
```

KnightMiner replied:
Just seems like an arbitrary change unless the method was removed. If this was a method removal that prompted it then its fine though it seems like something I would not expect to be removed.

As an aside, two empty stacks would have matching items due to the way `getItem()` works. So effectively the is empty check means that two empty stacks are not considered to have the same item despite both being empty.

queengooborg replied:
This change was made due to a method removal, yes -- `sameItem()` was replaced with `isSameItem()`, but since `isSameItem()` just calls `is()`, I opted to use `is()` directly!

ToCraft committed:
fix ElementScreens

ToCraft committed:
use RecipeCategory.MISC for now

queengooborg committed:
Use new method call

queengooborg committed:
ModelBakery bakery -> ModelBaker baker

KnightMiner commented on build.gradle:
```diff
@@ -221,7 +223,7 @@ dependencies {
     compileOnly fg.deobf("curse.maven:diet-443570:4813890")
 
-    compileOnly fg.deobf("dev.gigaherz.jsonthings:JsonThings-${minecraft_version}:${json_things_version}")
+    compileOnly fg.deobf("curse.maven:json-things-541617:5593943")
```
Why are you using curse maven? Json Things still has a proper maven on 1.20.1

queengooborg replied:
My guess is that it is because Json Things 0.9.2 (the most recent version) isn't available on the Maven repository, though that is only speculation on my part.  I just pushed a commit to switch back to using the Maven repo and use 0.9.1 instead!

ToCraft replied:
Yep, I couldn't find the newest version on their maven. But now we are using it again.

KnightMiner replied:
Is there a reason you need the newest version?

KnightMiner commented on build.gradle:
```diff
@@ -380,9 +382,11 @@ publishing {
 
 tasks.withType(JavaCompile).configureEach {
     options.encoding = 'UTF-8' // Use the UTF-8 charset for Java compilation
+    options.compilerArgs << "-Xmaxerrs" << "1000" // Show way more compiler errors
```
This should be an IDE run configuration change, not a global change to the build.gradle.

queengooborg replied:
Just to clarify, are you referring to only a _local_ configuration change?

queengooborg replied:
Just to clarify, are you referring to only a _local_ configuration change?

KnightMiner replied:
You can change your run configurations directly in IDEA settings. This is just the config that generates it.

That option does not need to be set when not porting.

ToCraft replied:
I changed it because I prefer compiling via Terminal instead of IDEA. But since we don't need that option anymore, I'll remove it.

queengooborg committed:
Switch back to using Json Things from Maven repo

KnightMiner commented on src/main/java/slimeknights/tconstruct/TConstruct.java:
```diff
@@ -225,27 +229,30 @@ private static void missingMappings(MissingMappingsEvent event) {
 
   /**
    * Gets a resource location for Tinkers
-   * @param name  Resource path
-   * @return  Location for tinkers
+   *
+   * @param name Resource path
+   * @return Location for tinkers
```
Don't make arbitrary reformatting changes. They just clutter what needs to be reviewed.

queengooborg replied:
These formatting changes appear to have been all created by the IntelliJ formatter (or Eclipse formatter, but I use IDEA CE and it is making the exact same suggestions), including the `<p>` addition further down below in this file.  I pushed a commit to undo them for the time being as it does not directly relate to 1.20 porting, but it would probably be good to run a formatter in the future!

KnightMiner replied:
That formatter has not been used in years.

In any case, it does not matter if the formatter is used or not, when making a PR don't make changes to stuff out of scope of the PR.

queengooborg committed:
Un-format JavaDoc comments

queengooborg committed:
Various fixes

queengooborg committed:
assemble() needs registry access

queengooborg committed:
renderButton -> renderWidget

ToCraft committed:
remove compiler configuration for a longer error list

ToCraft committed:
fix TinkerSmeltery

ToCraft committed:
fix Vector3f rotations

ToCraft committed:
port font.draw to graphics.drawString

queengooborg committed:
Optimize imports

queengooborg committed:
Fix a few graphics blits

queengooborg committed:
Bump mod version

queengooborg committed:
Fix one more font renderer

queengooborg committed:
Add required overrides

queengooborg committed:
Refactor ArmorSlotType to integrate built-in armor type enum

queengooborg committed:
Register particle factories

queengooborg committed:
Add missing variable

queengooborg committed:
Convert GadgetRecipeProvider

queengooborg committed:
Fix a quaternion transformation

queengooborg committed:
PoseStack -> GuiGraphics

queengooborg committed:
Various fixes

queengooborg committed:
Use DamageTypeTags to determine damage source types

queengooborg committed:
Add tags() as ProtectionModule builder method

Creates DamageSourcePredicate objects on the fly, following the removal of pre-existing objects in Mantle

queengooborg committed:
Various fixes

queengooborg committed:
Convert additional damage source types

queengooborg committed:
Optimize imports

ToCraft committed:
fix recipe.assemble

ToCraft committed:
fix LootContext.Builder

ToCraft committed:
fix getMaterials

queengooborg committed:
getMaterials() -> resolveParents()

queengooborg committed:
Replace Materials with BlockProperties

queengooborg committed:
getMaterials() -> resolveParents()

queengooborg committed:
Finish converting creative mode tabs

spppacecat commented:
the finishing is armour trims ?

queengooborg committed:
Various fixes

queengooborg committed:
Remove unneeded functions

queengooborg committed:
Various fixes

queengooborg committed:
Fix typo

queengooborg committed:
Fix typo (again)

KnightMiner commented:
> the finishing is armour trims ?

While armor trims are a feature desired in 1.20, its out of scope in a porting PR so should be left out. We have systems in place that will let us support them on our models, though that requires me comparing our models to vanilla changes to decide the best approach.

ToCraft committed:
fix renderItem

ToCraft committed:
fix Holder.hackyErase

queengooborg committed:
Fix ConfiguredFeature

queengooborg committed:
Disable fillItemCategory() methods

They seem to be no longer needed now, as they are a part of the old creative mode tab system.  There's no mention about them in the porting guides however, so they're simply commented out for easy readding.

queengooborg committed:
Fix recipe serializers

queengooborg committed:
Remove setSendRepeatsToGui()

queengooborg committed:
No need to tag non-flammable wood items, just flammable logs

queengooborg committed:
Fix renderTooltip() calls

queengooborg committed:
Add block lookup

queengooborg committed:
Make methods asynchronous as needed

queengooborg committed:
Pass in registry access

queengooborg committed:
FlexPartCastItem must implement IFlexItem

queengooborg committed:
Fix overrides

queengooborg committed:
REPLACEABLE_PLANTS -> REPLACEABLE

queengooborg committed:
Fix SlimeTreeFeature.isTallPlantAt

queengooborg committed:
Fix BlockBehaviour creation

queengooborg committed:
Update SlimeFungusBlock() constructors

queengooborg committed:
Ensure that replaceableBlocks is passed into SlimeFungusConfig

queengooborg committed:
Convert WorldgenDatapackRegistryProvider

ToCraft committed:
fix FlexBlockTypes

ToCraft committed:
fix Exception handling in ModelSpriteProvider

ToCraft committed:
fix RetexturedBlockItem.getTexture

ToCraft committed:
fix addSprite method

ToCraft committed:
fix ARMOR_MODIFIER_UUID_PER_TYPE not found

ToCraft committed:
remove unused StructureUpdater

ToCraft committed:
remove getMaterials method (was probably useless)

ToCraft committed:
fix AbstractStructureRepalleter Exception handling

ToCraft committed:
move DamageSources to DamageTypes

ToCraft committed:
small constructor fixes

ToCraft committed:
fix test sources

ToCraft committed:
code cleanup

Signed-off-by: To_Craft <development@tocraft.dev>

ToCraft committed:
add damage type tags

Signed-off-by: To_Craft <development@tocraft.dev>

Leclowndu93150 commented:
it's getting closer and closer to being able to load the game

furtabs commented:
> it's getting closer and closer to being able to load the game

i will say this if you arent following this project it already compiles which is really good, just need to load in game and finally not crash lol.

Love everyone participation on this. Keep it up

ToCraft commented:
@KnightMiner Could you take a look at the InventoryBlock in Mantle? Currently, it seems like it is missing `createBlockStateDefinition`.

<details>
<summary>Stacktrace <i>(Click to expand)</i></summary>

        Caused by: java.lang.IllegalArgumentException: Cannot get property EnumProperty{name=axis, clazz=class net.minecraft.core.Direction$Axis, values=[x, y, z]} as it does not exist in Block{minecraft:air}
                at TRANSFORMER/minecraft@1.20.1/net.minecraft.world.level.block.state.StateHolder.getValue(StateHolder.java:98)
                at TRANSFORMER/minecraft@1.20.1/net.minecraft.world.level.block.Blocks.lambda$log$51(Blocks.java:1173)
                at TRANSFORMER/minecraft@1.20.1/net.minecraft.world.level.block.state.BlockBehaviour$BlockStateBase.<init>(BlockBehaviour.java:474)
                at TRANSFORMER/minecraft@1.20.1/net.minecraft.world.level.block.state.BlockState.<init>(BlockState.java:14)
                at TRANSFORMER/minecraft@1.20.1/net.minecraft.world.level.block.state.StateDefinition.lambda$new$3(StateDefinition.java:61)
                at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
                at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
                at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
                at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)
                at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
                at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
                at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
                at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
                at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
                at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
                at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
                at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
                at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
                at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)
                at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
                at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
                at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
                at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
                at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
                at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
                at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
                at java.base/java.util.stream.Streams$StreamBuilderImpl.forEachRemaining(Streams.java:411)
                at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
                at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
                at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
                at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
                at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
                at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
                at TRANSFORMER/minecraft@1.20.1/net.minecraft.world.level.block.state.StateDefinition.<init>(StateDefinition.java:59)
                at TRANSFORMER/minecraft@1.20.1/net.minecraft.world.level.block.state.StateDefinition$Builder.create(StateDefinition.java:157)
                at TRANSFORMER/minecraft@1.20.1/net.minecraft.world.level.block.Block.<init>(Block.java:172)
                at TRANSFORMER/mantle@1.11/slimeknights.mantle.block.InventoryBlock.<init>(InventoryBlock.java:34)
                at TRANSFORMER/tconstruct@3.9.0.DEV.6f8e2a90e/slimeknights.tconstruct.shared.block.TableBlock.<init>(TableBlock.java:40)
                at TRANSFORMER/tconstruct@3.9.0.DEV.6f8e2a90e/slimeknights.tconstruct.tables.block.TabbedTableBlock.<init>(TabbedTableBlock.java:11)
                at TRANSFORMER/tconstruct@3.9.0.DEV.6f8e2a90e/slimeknights.tconstruct.tables.block.RetexturedTableBlock.<init>(RetexturedTableBlock.java:18)
                at TRANSFORMER/tconstruct@3.9.0.DEV.6f8e2a90e/slimeknights.tconstruct.tables.block.CraftingStationBlock.<init>(CraftingStationBlock.java:13)
                at TRANSFORMER/tconstruct@3.9.0.DEV.6f8e2a90e/slimeknights.tconstruct.tables.TinkerTables.lambda$static$2(TinkerTables.java:77)
                at TRANSFORMER/forge@47.3.12/net.minecraftforge.registries.DeferredRegister.lambda$addEntries$1(DeferredRegister.java:386)
                at TRANSFORMER/forge@47.3.12/net.minecraftforge.registries.RegisterEvent.register(RegisterEvent.java:59)
                at TRANSFORMER/forge@47.3.12/net.minecraftforge.registries.DeferredRegister.addEntries(DeferredRegister.java:386)
                at TRANSFORMER/forge@47.3.12/net.minecraftforge.registries.DeferredRegister$EventDispatcher.handleEvent(DeferredRegister.java:328)
                at TRANSFORMER/forge@47.3.12/net.minecraftforge.registries.__EventDispatcher_handleEvent_RegisterEvent.invoke(.dynamic)
                at MC-BOOTSTRAP/net.minecraftforge.eventbus/net.minecraftforge.eventbus.ASMEventHandler.invoke(ASMEventHandler.java:73)
                at MC-BOOTSTRAP/net.minecraftforge.eventbus/net.minecraftforge.eventbus.EventBus.post(EventBus.java:315)
                at MC-BOOTSTRAP/net.minecraftforge.eventbus/net.minecraftforge.eventbus.EventBus.post(EventBus.java:296)
                at LAYER PLUGIN/javafmllanguage@1.20.1-47.3.12/net.minecraftforge.fml.javafmlmod.FMLModContainer.acceptEvent(FMLModContainer.java:121)
                ... 32 more

</details>

EDIT: Nvm, I fixed it.

ToCraft committed:
fix TableBlock

Signed-off-by: To_Craft <development@tocraft.dev>

KnightMiner commented:
Let me be honest about where I see the state of things. This PR is not in a state where its less effort to review than it is to just port to 1.20 myself. This is most notably true as this PR was started nearly 4 months ago and the 1.19 branch has had a lot of changes since then, notably changes meant to prepare for 1.20. It shouldn't take 134 commits to port the mod to 1.20; I've done quite a few major ports before and none of them ended up close to that volume before the game even ran.

I also keep seeing commits in my notifications that just make me ask "why are you doing that?"; I feel there have been a lot of unneeded changes to the code base simply because you don't understand my codebase. Its possible some of these actually are required, but there are notable examples that make me less convinced. The most recent commit is a good example, there is no justification of why `TableBlock` now needs an axis property; unless its a change to `SimpleWaterloggedBlock` (which according to my sources it is not), that is just an addition to solve a crash without understanding why the crash was happening; you really should be asking why an axis property is being required in the first place. Another example is the removal of the "unused" datagenerator to update structure NBT; that data generator is used once per update so is kept around as we need it again after this update is done.

I recently have been pushing to `nobuild-api-break`, which is a (currently) 1.19 branch meant to give me an opportunity to clean up now unneeded code before the port to 1.20. There are notably a few APIs that we are planning to delete so there is no sense spending the effort to port them just to delete them. Once I'm done with that cleanup; *then* will be a good time to *start* the 1.20 port. Getting your port in sync with that branch just means even more changes I have to review.

Overall, like I said back when you asked me several months ago; its much more useful if you want to help with the port to work on specific tasks related to porting. I had a couple people help with that, but mostly I ended up having to do them myself. Now that those are done, I can work on the task that I always needed to do myself, which is getting the codebase to compile/run on 1.20. I have some time upcoming for the holiday break, which was when I always planned to do the port anyways.

So overall, I don't see this PR getting merged, the volume of changes you made unrelated to porting combined with the fact you still don't have a bootable game means its only going to delay the port further if I try to use your PR. Hopefully you did not find the effort a total waste as you learned something about making mods in the process.

KnightMiner commented:
> The most recent commit is a good example, there is no justification of why TableBlock now needs an axis property; unless its a change to SimpleWaterloggedBlock (which according to my sources it is not), that is just an addition to solve a crash without understanding why the crash was happening; you really should be asking why an axis property is being required in the first place. 

Hint: this is because you decided to make all tables copy their block properties from the oak log. Absolutely no idea why you decided to do that; my code in 1.19 never did that. One of those properties almost certainly makes use of axis.

If this is about the removal of materials, yeah, copying from a block is absolutely the wrong approach. Just use the proper builder calls to create the desired properties from the material.

Don't just stop a crash from happening. Ask why your fix is the right fix.

Leclowndu93150 commented:
ngl the codebase was altered very  badly in the PR

ToCraft commented:
At first, let me say that we used the [Mod Migration Primer](https://github.com/neoforged/.github/blob/main/primers/1.20/index.md) as reference for the porting, so most changes we did were actually required.

Moreover, I actually asked you multiple times for how you wanted some things to be ported. Since you didn't answer in time and not only ignored most of these requests in terms of content but also wanted us to stay as close to the 1.19 code as possible, we were forced to port with many workarounds that definitely weren't as elegant as they could have been when you would have accepted some basic background things to be improved and changed along the way.

Sadly, you constantly seemed to want us to stay with the old codebase and just alter it. That's not how porting can efficiently be done, so there was no way except to make some workarounds.

Regarding the latest change with the TableBlocks- why do you make such a great deal about it? You could have just said something when we tried to migrate from Materials to BlockBehaviours.
You also didn't stay productive in your comments. Instead of informative statements, we just received comments from your side similar to the following:

> Hint: this is because you decided to make all tables copy their block properties from the oak log. Absolutely no idea why you decided to do that; my code in 1.19 never did that. One of those properties almost certainly makes use of axis.

Doesn't that sound a bit derogatory towards us and our work? Even **if** it could have been of importance, you only supplied it *after* we wrote a working fix. Therefore, you probably knew it couldn't affect the port in any changing way. Instead, you could have just focused on a more acute error and actually contribute to the port.

Having this kind of feedback from your side, we don't see any way nor reason to work with you on this project.

I will now close this PR unfinished, as I do not expect you to have any interest in it.

Please note that we will not be following the comments in this PR any further.

BamsyTheSergal commented:
"my way or the highway" ahh moment

KnightMiner commented:
> At first, let me say that we used the [Mod Migration Primer](https://github.com/neoforged/.github/blob/main/primers/1.20/index.md) as reference for the porting, so most changes we did were actually required.

I gave several examples of changes that were not required. Those changes make me not trust the rest of the changes you made.

> Moreover, I actually asked you multiple times for how you wanted some things to be ported. Since you didn't answer in time and not only ignored most of these requests in terms of content but also wanted us to stay as close to the 1.19 code as possible, we were forced to port with many workarounds that definitely weren't as elegant as they could have been when you would have accepted some basic background things to be improved and changed along the way.

No, you really didn't. You asked me before starting the PR and I told you don't start the PR. Then you asked me twice during the PR via github, former I answered and latter you "solved on your own". There was no other communication I received.

> Sadly, you constantly seemed to want us to stay with the old codebase and just alter it. That's not how porting can efficiently be done, so there was no way except to make some workarounds.

I've ported my mod more times that you have. You don't port a mod by completely rewriting everything. You port a mod by changing ***just what the port requires being changed***. I've already ported Mantle to 1.20, along with worked with someone doing a PR for one of my smaller mods. While Tinkers is a large mod, the amount of changes you are making is completely insane for porting; you are not porting, you are completely rewriting the mod. That is unreviewable.

> Regarding the latest change with the TableBlocks- why do you make such a great deal about it? You could have just said something when we tried to migrate from Materials to BlockBehaviours.
You also didn't stay productive in your comments. Instead of informative statements, we just received comments from your side similar to the following:

Because you pinged me about it then fixed it in the clearly wrong way. I have not had time to keep up with every commit you pushed, my plan was to review it at the end. But the longer you went, the more clear it was this PR is not producing a properly ported mod. 


> Doesn't that sound a bit derogatory towards us and our work? Even if it could have been of importance, you only supplied it after we wrote a working fix. Therefore, you probably knew it couldn't affect the port in any changing way. Instead, you could have just focused on a more acute error and actually contribute to the port.

So basically, you expect me to babysit you guys as you make the port to ensure every single change you make is the right change? Again, porting PRs only work if its less work for me than just porting myself; if I am babysitting you guys to make sure every fix is correct then there is no benefit to your PR.

I did not "know" the wood thing was wrong when you first did it because I am not watching every change. I just saw a clearly wrong change in your most recent commit, then spent less than 5 minutes figuring out why it was wrong.

> Having this kind of feedback from your side, we don't see any way nor reason to work with you on this project.
>
> I will now close this PR unfinished, as I do not expect you to have any interest in it.

You were not working with me before. I told you before you started what working with me looked like, and you refused. This project for the entire 4 months has been you ignoring my input and moving forward despite my suggestions.

I had several other people open PRs based on what I told you before you started this project which were helpful and reviewable. IF you really wanted to help you would have done the same.


> Please note that we will not be following the comments in this PR any further.

In other words, "I'm right and refuse to be told otherwise". Yeah, ignoring the wishes of the owner of a repo and then making a PR that multiple people say is being done the wrong way means I'm to blame. I guess I was wrong when I said you hopefully learned something from this, seems you did not learn anything.

KnightMiner commented:
> "my way or the highway" ahh moment

Not sure if you are talking about me or ToCraft. I will respond though that if you are making a PR or otherwise ***taking someone else's code as the core of your project***, its completely reasonable that they (that is, me in this case) get say over what is done with it. I've poured years of my life into this project, and I don't take kindly to people trying to take my work as their own when I'm still actively maintaining this mod. If I had abandoned this project with a public statement that would be different, but when the last commit on this project was literally yesterday, its not dead enough to need a fork.

Since a fork is wrong, PRs are the proper way to help the project. And yes, with PRs I am the repo owner, so I decide what gets merged and what does not get merged. If the author of a PR refuses to work with me, then the PR will never be merged.

----

To anyone discovering this PR in the future, let me reiterate on a few things.

I am porting the mod to 1.20. I've already started the port, I've just not made all the work public. I don't need someone to compile the mod against 1.20 and fix compiler errors in a PR; that ends up being more work to review than for me to just do it myself. I told the author of this PR that before they started but they did not care.

There are plenty of other tasks you can help with if you really want to help port the mod. Join our discord and ask in the `#coding` channel if you are interested. Note that porting time tends to be one of the hardest times to help though, as the majority of port itself is very inefficient to split between multiple people, which I think this PR just proves. There will probably be a bunch of smaller tasks to do once I have a working 1.20 branch, notably armor trim support and other new 1.20 features.

BamsyTheSergal commented:
> Not sure if you are talking about me or ToCraft.

i was talking about ToCraft's blatant ignorant takes. You guys do a lot of work on this stuff. so i appreciate all of it. i generally considered learning java properly just to help you guys out any way i can, i have been a very long time user of tinkers and fully enjoy any change that happens, but life gets in the way.

MageOfGaming commented:
So since I was kinda the person who started all of this on the DC a few months ago I kinda feel obligated to say this:

I could or we could go over so many stupid takes, simply misinformation and weird opinions in the entire thread of this PR, there is honestly a lot to go over and it hits badly in both directions, from a social and technical perspective, now since this PR was official closed there is no need to do this however and would simply lead to more stress for all of us, I've watched this PR since it was created and by now I simply feel sick to the stomach getting the notification about it since most discussions in here have slowly acquired this passive aggressive style to it that is just incredibly toxic, in general the discussion and communication here was incredibly toxic (Which is one of the reasons why I never commented on this earlier, I definitely had things to say, I decided not to however, to not get involved once more with this ****show of a "discussion"), I won't call out anyone specific, the people I am addressing know that I'm doing so

So why do I write this?

Like I just explained this PR is closed and was filled with stressful and toxic communication everywhere, so please just move on, sparking another discussion after the PR has already been closed is so unnecessary, I can understand that people feel offended and then reply because of that but that won't bring you guys anywhere from here on, especially since To_craft isn't going to respond and I'm not going to discuss this further either (Otherwise I wouldn't appeal for this thread to just rest and never come up again unless it's VERY necessary) 

So honestly, really just leave this be, this goes for anyone seeing this in the future as well, responding any further from here on will just cause stress for a lot of people.
